cc/dual_net/tf_mgpu_dual_net.cc

```diff
@@ -0,0 +1,199 @@
+      auto* feature_tensor = inputs_.front().second.flat<float>().data();
+      // Copy the features into the input tensor.
+      for (const auto* feature : features) {
+        feature_tensor =
+            std::copy(feature->begin(), feature->end(), feature_tensor);
+      }
+
+      // Run the model.
+      TF_CHECK_OK(session_->Run(inputs_, output_names_, {}, &outputs_));
+
+      // Copy the policy and value out of the output tensors.
+      const auto& policy_tensor = outputs_[0].flat<float>();
+      const auto& value_tensor = outputs_[1].flat<float>();
+      for (int i = 0; i < batch_size; ++i) {
+        memcpy(outputs[i]->policy.data(), policy_tensor.data() + i * kNumMoves,
+               sizeof(outputs[i]->policy));
+        outputs[i]->value = value_tensor.data()[i];
+      }
+    }
+
+   private:
+    std::unique_ptr<tensorflow::Session> session_;
+    std::vector<std::pair<std::string, tensorflow::Tensor>> inputs_;
+    std::vector<std::string> output_names_;
+    std::vector<tensorflow::Tensor> outputs_;
+  };
+
+  struct InferenceData {
+    std::vector<const DualNet::BoardFeatures*> features;
+    std::vector<DualNet::Output*> outputs;
+    absl::Notification* notification;
+  };
+
+ public:
+  explicit TfMgpuDualNet(std::string graph_path);
+
+  ~TfMgpuDualNet() override;
+
+  void RunMany(std::vector<const BoardFeatures*> features,
+               std::vector<Output*> outputs, std::string* model) override;
+
+ private:
+  static void PlaceOnDevice(tensorflow::GraphDef* graph_def,
+                            const std::string& device) {
+    for (auto& node : *graph_def->mutable_node()) {
+      if (node.op() == "Const") {
+        auto it = node.attr().find("dtype");
+        if (it != node.attr().end() &&
+            it->second.type() == tensorflow::DT_INT32) {
+          continue;  // Const nodes of type int32 need to be in CPU.
+        }
+      }
+      node.set_device(device);
+    }
+  }
+
+  std::string graph_path_;
+  ThreadSafeQueue<InferenceData> inference_queue_;
+  std::vector<std::thread> worker_threads_;
+  std::atomic<bool> running_;
+};
+
+TfMgpuDualNet::TfMgpuDualNet(std::string graph_path)
+    : graph_path_(graph_path), running_(true) {
+  GraphDef graph_def;
+
+  // If we can't find the specified graph, try adding a .pb extension.
+  auto* env = Env::Default();
+  if (!env->FileExists(graph_path).ok()) {
+    graph_path = absl::StrCat(graph_path, ".pb");
+  }
+
+  TF_CHECK_OK(ReadBinaryProto(env, graph_path, &graph_def));
+
+  auto functor = [this](const tensorflow::GraphDef& graph_def) {
+    TfWorker worker(graph_def);
+    while (running_) {
+      InferenceData inference;
+      if (inference_queue_.PopWithTimeout(&inference, absl::Seconds(1))) {
+        worker.RunMany(std::move(inference.features),
+                       std::move(inference.outputs));
+        inference.notification->Notify();
+      }
+    }
+  };
+
+  TF_CHECK_OK(tensorflow::ValidateGPUMachineManager());
```

Comment:
Sorry about breaking the GPU-less TensorFlow path! I assume that's what happened, because the single-GPU path should not be affected. 

For the GPU-less path, I think it would be sufficient to add a single worker thread here that's not placed on a device. Something like replacing line 162 with:

if (!tensorflow::ValidateGpuMachineManager().ok()) {
  worker_threads_.emplace_back(functor, graph_def);
  return;
}

Reply:
Yes, it was the ability to run on the CPU that I noticed; I haven't tried the single-GPU path.

If I take the approach you describe of having a single worker that's not on a device, I'll still have the need to `#ifdef`-out GPU-related code, correct? It was a missing `#include` that led me down this path in the first place.

Reply:
Tested #493 and it worked great for me. Will withdraw this request.